### PR TITLE
Renew auth before end of token

### PR DESCRIPTION
**Pull request by Errorname**
This PR changes the authentication workflow:

* When a user loads the application:
  * If first time or new computer: show login page which redirects to Auth0 login page (3 clicks)
  * Else: renew token in the background while showing a loader (0 clicks)
* While a user is authenticated:
  * 5min before the end of its token validity: renew token in the background without showing a loader

### CODE REVIEW

File: src/services/auth.js [hunk 1]
```diff
@@ -10,85 +10,104 @@ class Auth {
...
-            reject(new Error('Authentication did not work'))
-          }
+      this.auth0.parseHash({ hash }, (err, authResult) => {
+        if (authResult && authResult.accessToken && authResult.idToken) {
+          resolve(authResult)
+        } else if (err) {
+          alert(`Error: ${err.error}. Check the console for further details.`)
+          reject(err)
+        } else {
+          reject(new Error('Authentication did not work'))
```
Comment (hgwood) [inline src/services/auth.js]: Actually this case is not an error, it means there is no hash to parse.

File: src/services/auth.js [hunk 2]
```diff
@@ -10,85 +10,104 @@ class Auth {
...
-    const accessToken = this.session ? this.session.accessToken : null
-    if (!accessToken) {
-      throw new Error('No access token found')
-    }
-    return accessToken
+  /* Internal actions methods */
+
+  renewAuth = () => {
+    return new Promise((resolve, reject) => {
+      this.auth0.checkSession({}, (err, authResult) => {
```
Comment (hgwood) [inline src/services/auth.js]: Nice! I need to use this too.

File: src/services/auth.js [hunk 3]
```diff
@@ -97,15 +116,42 @@ class Auth {
...
-    const url = sessionStorage.getItem('after_login_redirect_url')
+  wasAuthenticated () {
+    return (
+      this.session &&
+      this.session.expiresAt < new Date().getTime() &&
+      localStorage.userId
+    )
+  }
+
+  getAfterLoginRedirectUrl () {
```
Comment (hgwood) [inline src/services/auth.js]: It seems a bit misleading to have a function named `get` that removes the thing after getting it. Maybe `pop` would be a better name (`popAfterLoginRedirectLogin`). What do you think?

File: src/services/auth.js [hunk 4]
```diff
@@ -97,15 +116,42 @@ class Auth {
...
+    const url = sessionStorage.after_login_redirect_url
     sessionStorage.removeItem('after_login_redirect_url')
     return url
   }
+
+  setUserId (id) {
+    localStorage.userId = id
+  }
+
+  getUserId () {
```
Comment (hgwood) [inline src/services/auth.js]: Same comment as for `getAfterLoginRedirectUrl`.

File: src/services/auth.js [hunk 5]
```diff
@@ -10,85 +10,104 @@ class Auth {
...
-            alert(`Error: ${err.error}. Check the console for further details.`)
-            reject(err)
-          } else {
-            reject(new Error('Authentication did not work'))
-          }
+      this.auth0.parseHash({ hash }, (err, authResult) => {
+        if (authResult && authResult.accessToken && authResult.idToken) {
+          resolve(authResult)
+        } else if (err) {
+          alert(`Error: ${err.error}. Check the console for further details.`)
```
Comment (hgwood) [inline src/services/auth.js]: Not quite the role of the authentication code to directly access the user visual space. Let the app (the caller) decide what to show the user.

File: src/services/auth.js (hunk 1 above)
Comment (Errorname) [reply src/services/auth.js]: Oh ? I couldn't find documentation about this

File: src/services/auth.js (hunk 3 above)
Comment (Errorname) [reply src/services/auth.js]: Agreed, this would give more context about what it does

File: src/services/auth.js (hunk 1 above)
Comment (hgwood) [reply src/services/auth.js]: I have empirical evidence ^^

**Pull request merged**